src/rgw/rgw_op.cc

```diff
@@ -1100,27 +1108,67 @@ int RGWGetObj::verify_permission(optional_yield y)
     if (has_s3_existing_tag || has_s3_resource_tag)
       rgw_iam_add_objtags(this, s, has_s3_existing_tag, has_s3_resource_tag);
 
-  if (get_torrent) {
-    if (s->object->get_instance().empty()) {
-      action = rgw::IAM::s3GetObjectTorrent;
-    } else {
-      action = rgw::IAM::s3GetObjectVersionTorrent;
+  // if it's system request and wrapped then we assume rgwx-uid was passed by the request and so it's a replication request.
+  bool is_replication_request = s->system_request && s->auth.identity->is_system_wrapped();
+
+  if (s->bucket->get_info().obj_lock_enabled()) {
+    get_retention = verify_object_permission(this, s, rgw::IAM::s3GetObjectRetention);
+    if (is_replication_request && !get_retention) {
+      set_retention_status_header(this, s->object.get(), y, "FAILED");
+
+      return -EACCES;
     }
-  } else {
-    if (s->object->get_instance().empty()) {
-      action = rgw::IAM::s3GetObject;
-    } else {
-      action = rgw::IAM::s3GetObjectVersion;
+
+    get_legal_hold = verify_object_permission(this, s, rgw::IAM::s3GetObjectLegalHold);
+    if (is_replication_request && !get_legal_hold) {
+      set_legal_hold_status_header(this, s->object.get(), y, "FAILED");
+
+      return -EACCES;
     }
   }
 
-  if (!verify_object_permission(this, s, action)) {
-    return -EACCES;
+  if (is_replication_request) {
+    // check for s3:GetObject(Version)Acl permission
+    action = s->object->get_instance().empty() ? rgw::IAM::s3GetObjectAcl : rgw::IAM::s3GetObjectVersionAcl;
+    if (!verify_object_permission(this, s, action)) {
+      // update the replication status of the object to FAILED
+      set_replication_status_header(this, s->object.get(), y, "FAILED");
+
+      return -EACCES;
+    }
+
+    // check for s3:GetObject(Version)ForReplication permission
+    action = s->object->get_instance().empty() ? rgw::IAM::s3GetObjectForReplication : rgw::IAM::s3GetObjectVersionForReplication;
```

Comment:
what do you think about using `s3GetObjectVersionForReplication` unconditionally here, to avoid the need for an rgw-specific `s3GetObjectForReplication`?

if the source bucket is versioned, i'd expect all sync requests to include a versionId. and if non-versioned, none of the sync requests would. can you think of any cases where the policy would need to deny one but allow the other?

Reply:
The only case would be someone suspending the versioning and expecting the policy to stop replicating old versioned objects. I guess after suspension `s->object->get_instance().empty()` would result in true, right?

Reply:
i ran `test_version_suspended_incremental_sync` and checked the radosgw logs - i see `REQUEST_URI=/saxpqz-1/obj?versionId=null` for those sync requests